File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CoursePlagiarismLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.R

/** Course plagiarism / AI-authorship settings helpers (M13.7). */
object CoursePlagiarismLogic {
    data class ProviderOption(
        val value: String,
        val labelRes: Int,
    )

    data class FormDraft(
        val checksEnabled: Boolean = true,
        val provider: String = "",
        val thresholdPct: String = "40",
    )

    enum class ValidationError {
        ThresholdInvalid,
    }

    const val DEFAULT_THRESHOLD_PCT = 40.0

    val providerOptions: List<ProviderOption> = listOf(
        ProviderOption("", R.string.mobile_courseSettings_plagiarism_provider_default),
        ProviderOption("none", R.string.mobile_courseSettings_plagiarism_provider_none),
        ProviderOption("turnitin", R.string.mobile_courseSettings_plagiarism_provider_turnitin),
        ProviderOption("copyleaks", R.string.mobile_courseSettings_plagiarism_provider_copyleaks),
        ProviderOption("gptzero", R.string.mobile_courseSettings_plagiarism_provider_gptzero),
    )

    fun cacheKey(courseCode: String): String = "course:$courseCode:plagiarism-settings"

    fun saveIdempotencyKey(courseCode: String): String = "course-plagiarism:$courseCode:save"

    fun patchPath(courseCode: String): String = "/api/v1/courses/$courseCode/plagiarism-settings"

    fun draft(settings: CoursePlagiarismSettings?): FormDraft =
        FormDraft(
            checksEnabled = settings?.plagiarismChecksEnabled ?: true,
            provider = normalizedProvider(settings?.plagiarismProvider),
            thresholdPct = formatThreshold(settings?.plagiarismAlertThresholdPct),
        )

    fun isDirty(current: FormDraft, baseline: FormDraft): Boolean = current != baseline

    fun validateDraft(draft: FormDraft): ValidationError? =
        if (parsedThreshold(draft.thresholdPct) == null) ValidationError.ThresholdInvalid else null

    fun buildPatchBody(current: FormDraft): PatchCoursePlagiarismBody =
        PatchCoursePlagiarismBody(
            plagiarismChecksEnabled = current.checksEnabled,
            plagiarismProvider = current.provider.takeIf { it.isNotEmpty() },
            plagiarismAlertThresholdPct = parsedThreshold(current.thresholdPct) ?: DEFAULT_THRESHOLD_PCT,
        )

    fun normalizedProvider(provider: String?): String {
        val trimmed = provider.orEmpty().trim().lowercase()
        return providerOptions.firstOrNull { it.value == trimmed && it.value.isNotEmpty() }?.value.orEmpty()
    }

    fun formatThreshold(value: Double?): String {
        val resolved = value?.takeIf { it.isFinite() } ?: DEFAULT_THRESHOLD_PCT
        return if (resolved % 1.0 == 0.0) resolved.toInt().toString() else resolved.toString()
    }

    fun parsedThreshold(text: String): Double? {
        val value = text.trim().toDoubleOrNull() ?: return null
        if (!value.isFinite() || value < 0 || value > 100) return null
        return value
    }
}

File: clients/ios/Lextures/Core/LMS/CoursePlagiarismLogic.swift
import Foundation

/// Course plagiarism / AI-authorship settings helpers (M13.7).
enum CoursePlagiarismLogic {
    struct ProviderOption: Identifiable, Hashable {
        var id: String { value }
        var value: String
        var labelKey: String
    }

    struct FormDraft: Equatable, Hashable {
        var checksEnabled: Bool
        var provider: String
        var thresholdPct: String
    }

    enum ValidationError: Equatable {
        case thresholdInvalid
    }

    static let defaultThresholdPct = 40.0

    static let providerOptions: [ProviderOption] = [
        ProviderOption(value: "", labelKey: "mobile.courseSettings.plagiarism.provider.default"),
        ProviderOption(value: "none", labelKey: "mobile.courseSettings.plagiarism.provider.none"),
        ProviderOption(value: "turnitin", labelKey: "mobile.courseSettings.plagiarism.provider.turnitin"),
        ProviderOption(value: "copyleaks", labelKey: "mobile.courseSettings.plagiarism.provider.copyleaks"),
        ProviderOption(value: "gptzero", labelKey: "mobile.courseSettings.plagiarism.provider.gptzero"),
    ]

    static func cacheKey(courseCode: String) -> String {
        "course:\(courseCode):plagiarism-settings"
    }

    static func saveIdempotencyKey(courseCode: String) -> String {
        "course-plagiarism:\(courseCode):save"
    }

    static func patchPath(courseCode: String) -> String {
        "/api/v1/courses/\(courseCode)/plagiarism-settings"
    }

    static func draft(from settings: CoursePlagiarismSettings?) -> FormDraft {
        FormDraft(
            checksEnabled: settings?.plagiarismChecksEnabled ?? true,
            provider: normalizedProvider(settings?.plagiarismProvider),
            thresholdPct: formatThreshold(settings?.plagiarismAlertThresholdPct)
        )
    }

    static func isDirty(current: FormDraft, baseline: FormDraft) -> Bool {
        current != baseline
    }

    static func validateDraft(_ draft: FormDraft) -> ValidationError? {
        parsedThreshold(draft.thresholdPct) == nil ? .thresholdInvalid : nil
    }

    static func buildPatchBody(current: FormDraft) -> PatchCoursePlagiarismBody {
        PatchCoursePlagiarismBody(
            plagiarismChecksEnabled: current.checksEnabled,
            plagiarismProvider: current.provider.isEmpty ? nil : current.provider,
            plagiarismAlertThresholdPct: parsedThreshold(current.thresholdPct) ?? defaultThresholdPct
        )
    }

    static func normalizedProvider(_ provider: String?) -> String {
        let trimmed = (provider ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard providerOptions.contains(where: { $0.value == trimmed && !$0.value.isEmpty }) else {
            return ""
        }
        return trimmed
    }

    static func formatThreshold(_ value: Double?) -> String {
        let resolved = value.flatMap { $0.isFinite ? $0 : nil } ?? defaultThresholdPct
        if resolved.rounded() == resolved {
            return String(Int(resolved))
        }
        return String(resolved)
    }

    static func parsedThreshold(_ text: String) -> Double? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let value = Double(trimmed), value.isFinite, value >= 0, value <= 100 else {
            return nil
        }
        return value
    }
}
